internal/handler/command.go

```diff
@@ -0,0 +1,420 @@
+
+	if strings.ToLower(method) == "get" {
+		return execReadCmd(&d, cmd)
+	} else {
+		appErr := execWriteCmd(&d, cmd, body)
+		return nil, appErr
+	}
+}
+
+func execReadCmd(device *models.Device, cmd string) (*models.Event, common.AppError) {
+	readings := make([]models.Reading, 0, common.CurrentConfig.Device.MaxCmdOps)
+
+	// make ResourceOperations
+	ros, err := cache.Profiles().ResourceOperations(device.Profile.Name, cmd, "get")
+	if err != nil {
+		common.LoggingClient.Error(err.Error())
+		return nil, common.NewNotFoundError(err.Error(), err)
+	}
+
+	if len(ros) > common.CurrentConfig.Device.MaxCmdOps {
+		msg := fmt.Sprintf("Handler - execReadCmd: MaxCmdOps (%d) execeeded for dev: %s cmd: %s method: GET",
+			common.CurrentConfig.Device.MaxCmdOps, device.Name, cmd)
+		common.LoggingClient.Error(msg)
+		return nil, common.NewServerError(msg, nil)
+	}
+
+	reqs := make([]ds_models.CommandRequest, len(ros))
+
+	for i, op := range ros {
+		objName := op.Object
+		common.LoggingClient.Debug(fmt.Sprintf("Handler - execReadCmd: deviceObject: %s", objName))
+
+		// TODO: add recursive support for resource command chaining. This occurs when a
+		// deviceprofile resource command operation references another resource command
+		// instead of a device resource (see BoschXDK for reference).
+
+		devObj, ok := cache.Profiles().DeviceObject(device.Profile.Name, objName)
+		common.LoggingClient.Debug(fmt.Sprintf("Handler - execReadCmd: deviceObject: %v", devObj))
+		if !ok {
+			msg := fmt.Sprintf("Handler - execReadCmd: no devobject: %s for dev: %s cmd: %s method: GET", objName, device.Name, cmd)
+			common.LoggingClient.Error(msg)
+			return nil, common.NewServerError(msg, nil)
+		}
+
+		reqs[i].RO = op
+		reqs[i].DeviceObject = devObj
+	}
+
+	results, err := common.Driver.HandleReadCommands(&device.Addressable, reqs)
+	if err != nil {
+		msg := fmt.Sprintf("Handler - execReadCmd: error for Device: %s cmd: %s, %v", device.Name, cmd, err)
+		return nil, common.NewServerError(msg, err)
+	}
+
+	var transformsOK bool = true
+
+	for _, cv := range results {
+		// get the device resource associated with the rsp.RO
+		do, ok := cache.Profiles().DeviceObject(device.Profile.Name, cv.RO.Object)
+		if !ok {
+			msg := fmt.Sprintf("Handler - execReadCmd: no devobject: %s for dev: %s in Command Result %v", cv.RO.Object, device.Name, cv)
+			common.LoggingClient.Error(msg)
+			return nil, common.NewServerError(msg, nil)
+		}
+
+		if common.CurrentConfig.Device.DataTransform {
+			err = transformer.TransformReadResult(cv, do.Properties.Value)
+			if err != nil {
+				common.LoggingClient.Error(fmt.Sprintf("Handler - execReadCmd: CommandValue (%s) transformed failed: %v", cv.String(), err))
+				transformsOK = false
+			}
+		}
+
+		err = transformer.CheckAssertion(cv, do.Properties.Value.Assertion, device)
+		if err != nil {
+			common.LoggingClient.Error(fmt.Sprintf("Handler - execReadCmd: Assertion failed for device resource: %s, with value: %s", cv.String(), err))
```

Comment:
Note we need to treat assertion failures in the command endpoint in the same way we handle them in async.go. Please open an issue for this...

Reply:
#92 opened for this issue

Reply:
#93 opened for the init cahce topic